src/hotspot/share/runtime/monitorDeflationThread.cpp

```diff
@@ -0,0 +1,98 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "classfile/javaClasses.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/java.hpp"
+#include "runtime/javaCalls.hpp"
+#include "runtime/monitorDeflationThread.hpp"
+#include "runtime/mutexLocker.hpp"
+
+MonitorDeflationThread* MonitorDeflationThread::_instance = NULL;
+
+void MonitorDeflationThread::initialize() {
+  EXCEPTION_MARK;
+
+  const char* name = "Monitor Deflation Thread";
+  Handle string = java_lang_String::create_from_str(name, CHECK);
+
+  // Initialize thread_oop to put it into the system threadGroup
+  Handle thread_group (THREAD, Universe::system_thread_group());
+  Handle thread_oop = JavaCalls::construct_new_instance(
+                          SystemDictionary::Thread_klass(),
+                          vmSymbols::threadgroup_string_void_signature(),
+                          thread_group,
+                          string,
+                          CHECK);
+
+  {
+    MutexLocker mu(THREAD, Threads_lock);
+    MonitorDeflationThread* thread =  new MonitorDeflationThread(&monitor_deflation_thread_entry);
+
+    // At this point it may be possible that no osthread was created for the
+    // JavaThread due to lack of memory. We would have to throw an exception
+    // in that case. However, since this must work and we do not allow
+    // exceptions anyway, check and abort if this fails.
+    if (thread == NULL || thread->osthread() == NULL) {
+      vm_exit_during_initialization("java.lang.OutOfMemoryError",
+                                    os::native_thread_creation_failed_msg());
+    }
+
+    java_lang_Thread::set_thread(thread_oop(), thread);
+    java_lang_Thread::set_priority(thread_oop(), NearMaxPriority);
+    java_lang_Thread::set_daemon(thread_oop());
+    thread->set_threadObj(thread_oop());
+    _instance = thread;
+
+    Threads::add(thread);
+    Thread::start(thread);
+  }
+}
+
+void MonitorDeflationThread::monitor_deflation_thread_entry(JavaThread* jt, TRAPS) {
+  while (true) {
+    {
+      // Need state transition ThreadBlockInVM so that this thread
+      // will be handled by safepoint correctly when this thread is
+      // notified at a safepoint.
+
+      // This ThreadBlockInVM object is not also considered to be
+      // suspend-equivalent because MonitorDeflationThread is not
+      // visible to external suspension.
+
+      ThreadBlockInVM tbivm(jt);
+
+      MonitorLocker ml(MonitorDeflation_lock, Mutex::_no_safepoint_check_flag);
+      while (!ObjectSynchronizer::is_async_deflation_needed()) {
+        // Wait until notified that there is some work to do.
+        // We wait for GuaranteedSafepointInterval so that
+        // is_async_deflation_needed() is checked at the same interval.
+        ml.wait(GuaranteedSafepointInterval);
```

Comment:
We may still decide to do this fix (even though the _object field is now weak):

    JDK-8249638 Re-instate idle monitor deflation as part of System.gc()
    https://bugs.openjdk.java.net/browse/JDK-8249638

and if we do, then we'll still need the request mechanism.

Reply:
So why not use a local semaphore and wait with safepoint check instead?

Reply:
Sorry my preference is for Monitors instead of semaphores. Let's
take that discussion off this PR and you can explain why you dislike
the Monitor so much and think the local semaphore is the way to go.

Reply:
Yes

Reply:
Filed the following new RFE:
    JDK-8256241 replace MonitorDeflation_lock with a semaphore
    https://bugs.openjdk.java.net/browse/JDK-8256241